src/runtime/pkg/device/drivers/utils.go

```diff
@@ -157,6 +160,92 @@ func checkIgnorePCIClass(pciClass string, deviceBDF string, bitmask uint64) (boo
 	return false, nil
 }
 
+func getMajorMinorFromDevPath(devPath string) (uint32, uint32, error) {
+	fi, err := os.Stat(devPath)
+	if err != nil {
+		return 0, 0, err
+	}
+
+	dev := fi.Sys().(*syscall.Stat_t)
+	return uint32(dev.Rdev >> 8), uint32(dev.Rdev & 0xff), nil
+}
+
+func extractIndex(devicePath string) (string, error) {
+
+	base := filepath.Base(devicePath)
+
+	const prefix = "vfio"
+	if !strings.HasPrefix(base, prefix) {
+		return "0", fmt.Errorf("unexpected device name format: %s", base)
+	}
+	return strings.TrimPrefix(base, prefix), nil
+}
+
+func getBdfFromVFIODev(major uint32, minor uint32) (string, error) {
+	devPath := fmt.Sprintf("/sys/dev/char/%d:%d", major, minor)
+	realPath, err := filepath.EvalSymlinks(devPath)
+	if err != nil {
+		return "", fmt.Errorf("Failed to resolve symlink for %s: %v", devPath, err)
+	}
+
+	bdfRegex := regexp.MustCompile(`([0-9a-fA-F]{4}:[0-9a-fA-F]{2}:[0-9a-fA-F]{2}\.[0-9a-fA-F])`)
+	matches := bdfRegex.FindAllString(realPath, -1)
+	if len(matches) == 0 {
+		return "", fmt.Errorf("No BDF found in resolved path: %s", realPath)
+	}
+	return matches[len(matches)-1], nil
```

Comment:
the regexp can be avoided by continuing to use `filepath`, e.g., `Clean()` and `Base()`.

Reply:
There are more steps needed to avoid the regexp, the symlink looks like this: 
```
 /sys/dev/char/506:0 -> ../../devices/pci0000:15/0000:15:01.0/0000:16:00.0/0000:17:02.0/0000:19:00.0/0000:1a:00.0/0000:1b:00.0/vfio-dev/vfio0
```
The device is `0000:1b:00.0/` not `vfio0`

Reply:
I misread the docs on `Clean()` but the basic idea I had was to drop `vfio-dev/vfio0` suffix and just take the `Base()` from the trimmed `realPath`

Reply:
@mythi Now without regex.